### PR TITLE
Support to specify LR of DCN's conv_offset

### PR DESCRIPTION
**Pull request by ZwwWayne**
This PR add an option `dcn_offset_lr_mult` in `params_cfg`. The modification enables to set the learning rate of DCN's conv offset layer to stabilize training.

### CODE REVIEW

File: mmcv/runner/optimizer/default_constructor.py [hunk 1]
```diff
@@ -28,6 +28,8 @@ class DefaultOptimizerConstructor(object):
       layers.
     - ``bypass_duplicate`` (bool): If true, the duplicate parameters
       would not be added into optimizer. Default: False.
+    - ``dcn_offset_lr_mult`` (float): It will be multiplied to the learning
+      rate for all parameters of offset layer in deformable convs.
```
Comment (hellock) [inline mmcv/runner/optimizer/default_constructor.py:32]: Move it before `bypass_duplicate`?

File: mmcv/runner/optimizer/default_constructor.py [hunk 2]
```diff
@@ -118,6 +121,11 @@ def add_params(self, params, module, prefix=''):
             # bias_lr_mult affects all bias parameters except for norm.bias
             if name == 'bias' and not is_norm:
                 param_group['lr'] = self.base_lr * bias_lr_mult
+
+            if prefix.find('conv_offset') != -1 and not is_norm:
```
Comment (hellock) [inline mmcv/runner/optimizer/default_constructor.py:125]: Is it safe? We may check the module type.

File: mmcv/runner/optimizer/default_constructor.py [hunk 3]
```diff
@@ -117,7 +124,7 @@ def _is_in(self, param_group, param_group_list):
 
         return not param.isdisjoint(param_set)
 
-    def add_params(self, params, module, prefix=''):
+    def add_params(self, params, module, prefix='', is_dcn_submodule=False):
```
Comment (xvjiarui) [inline mmcv/runner/optimizer/default_constructor.py:127]: Missing docstring.

File: mmcv/runner/optimizer/default_constructor.py [hunk 4]
```diff
@@ -182,14 +198,21 @@ def add_params(self, params, module, prefix=''):
...
                             'weight_decay'] = self.base_wd * dwconv_decay_mult
                     # bias lr and decay
-                    elif name == 'bias':
+                    elif name == 'bias' and not is_dcn_submodule:
+                        # TODO: current bias_decay_mult will have affect on DCN
                         param_group[
                             'weight_decay'] = self.base_wd * bias_decay_mult
             params.append(param_group)
 
+        from mmcv.ops import DeformConv2d
```
Comment (xvjiarui) [inline mmcv/runner/optimizer/default_constructor.py:207]: We may also add `ModulateDeformConv2d`.

File: mmcv/runner/optimizer/default_constructor.py [hunk 5]
```diff
@@ -167,10 +177,18 @@ def add_params(self, params, module, prefix=''):
...
+
             if not is_custom:
-                # bias_lr_mult affects all bias parameters except for norm.bias
-                if name == 'bias' and not is_norm:
+                # bias_lr_mult affects all bias parameters
+                # except for norm.bias dcn.conv_offset.bias
+                if name == 'bias' and not (is_norm or is_dcn_submodule):
                     param_group['lr'] = self.base_lr * bias_lr_mult
+
+                if (prefix.find('offset') != -1 and is_dcn_submodule
```
Comment (hellock) [inline mmcv/runner/optimizer/default_constructor.py:187]: This may not be safe. How about a branch named xxx_offset in some model and it adopts a DCN?

Comment (ZwwWayne) [reply mmcv/runner/optimizer/default_constructor.py:187]: How about we also pass an argument to DCN's conv_offset like `reset_lr` or so? If `reset_lr=True` and the module is a DCN's submodule, the LR is reset by the `dcn_offset_lr_mult`.

**Comment by hellock** [comment]
Task linked: [CU-49u01p Support to specify LR of DCN's conv_offset](https://app.clickup.com/t/49u01p)

**Review by xvjiarui — APPROVED** [review]

**Review by hellock — CHANGES_REQUESTED** [review]
Adding `offset_lr_mult` to the init method may be confusing when users are not using the default optimizer constructor.

File: mmcv/runner/optimizer/default_constructor.py [hunk 6]
```diff
@@ -27,19 +27,34 @@ class DefaultOptimizerConstructor:
...
+    - ``dcn_offset_lr_mult`` (float): It will be multiplied to the learning
+      rate for parameters of offset layer in the deformable convs
+      of a model.
     - ``bypass_duplicate`` (bool): If true, the duplicate parameters
       would not be added into optimizer. Default: False.
 
+    Note:
+        1. If the option ``dcn_offset_lr_mult`` is used, the constructor will
+            override the effect of ``bias_lr_mult`` in the bias of offset
+            layer. So be careful when use both ``bias_lr_mult`` and
```
Comment (hellock) [inline mmcv/runner/optimizer/default_constructor.py:49]: when using

File: mmcv/runner/optimizer/default_constructor.py [hunk 7]
```diff
@@ -117,7 +132,7 @@ def _is_in(self, param_group, param_group_list):
 
         return not param.isdisjoint(param_set)
 
-    def add_params(self, params, module, prefix=''):
+    def add_params(self, params, module, prefix='', is_dcn_submodule=None):
```
Comment (hellock) [inline mmcv/runner/optimizer/default_constructor.py:135]: `is_dcn_module` may be more straightforward.

File: mmcv/runner/optimizer/default_constructor.py [hunk 8]
```diff
@@ -182,14 +209,22 @@ def add_params(self, params, module, prefix=''):
...
                             'weight_decay'] = self.base_wd * dwconv_decay_mult
                     # bias lr and decay
-                    elif name == 'bias':
+                    elif name == 'bias' and not is_dcn_submodule:
+                        # TODO: current bias_decay_mult will have affect on DCN
                         param_group[
                             'weight_decay'] = self.base_wd * bias_decay_mult
             params.append(param_group)
 
+        from mmcv.ops import DeformConv2d, ModulatedDeformConv2d
```
Comment (hellock) [inline mmcv/runner/optimizer/default_constructor.py:218]: only `mmcv-full` contains these modules.

**Review by hellock — APPROVED** [review]
